### DISCUSSION TITLE
Problem importing pandas and numpy package

### DISCUSSION THREAD

**Discussion by NickatLCB** [cwapi3d/cwapi3dpython#66] [General] [answered]
Hello all.

I am attempting the pandas module into my script through the cadwork api. However, I am having issues with the NumPy dependency. It seems NumPy and my version of Python are in conflict. I created a conda environment with python 3.9 and versions of pandas and numpy that correspond, but still have issues. I have added the path to these packages to my sys.path variable.

Does anyone know what could be the problem?

**Comment by ccalvocm** [comment]
Greetings @NickatLCB 

Working fine in python=3.9.16 with anaconda. From terminal type:

1. `conda create -n cwapi python=3.9.16`
2. `conda activate cwapi`
3.  `pip install cwapi3d`

Kind regards

**Comment by roehligas** [comment, marked as answer]
I remember struggling with it once upon a time.

I solved it scanning the sys paths in my normal python environment and pretty much appending all of the necessary ones to my python scripts inside cadwork, see below.


```
# import modules
import  cadwork   
import  element_controller      as ec
import  attribute_controller    as ac
import geometry_controller      as gc
import utility_controller as uc
import visualization_controller as vc
import menu_controller as mc
import sys

# Adding the Anaconda Environment paths to the system paths
# This is necessary to be able to access external packages
sys.path.extend(['',
'K:\\02 KONSTRUKSJON\\04 CADWORK\\external Anaconda Environment\\Cadwork\\python39.zip', 
'K:\\02 KONSTRUKSJON\\04 CADWORK\\external Anaconda Environment\\Cadwork\\DLLs', 
'K:\\02 KONSTRUKSJON\\04 CADWORK\\external Anaconda Environment\\Cadwork\\lib', 
'K:\\02 KONSTRUKSJON\\04 CADWORK\\external Anaconda Environment\\Cadwork', 
'K:\\02 KONSTRUKSJON\\04 CADWORK\\external Anaconda Environment\\Cadwork\\lib\\site-packages'])


# Import external packages from Anaconda environment
import pandas as pd
```

**Comment by Brunner246** [comment]
The NumPy version that works within cadwork is: "1.22.2"

**Reply by roehligas** [reply]
I'm also running into trouble now.

Cadwork itself has no numpy on my pc.

![image](https://github.com/cwapi3d/cwapi3dpython/assets/66385500/a423d074-6f83-4235-813a-3459bad52928)

When i try to import numpy I get the following error:
`Traceback (most recent call last):
  File "//dc-1/Felles/02 KONSTRUKSJON/04 CADWORK/Cadwork_lib/USERPROFIL_30/3d/API.x64/ImportAttributesFromExcel/ImportAttributesFromExcel.py", line 69, in <module>
    import ImportAttributesFromExcel_V100 as tdo
  File "//dc-1/Felles\02 KONSTRUKSJON/04 CADWORK/Cadwork_lib/USERPROFIL_30/3d/API.x64/ImportAttributesFromExcel\ImportAttributesFromExcel_V100.py", line 29, in <module>
    import pandas as pd
  File "K:\02 KONSTRUKSJON\04 CADWORK\external Anaconda Environment\Cadwork\lib\site-packages\pandas\__init__.py", line 16, in <module>
    raise ImportError(
ImportError: Unable to import required dependencies:
numpy: 

IMPORTANT: PLEASE READ THIS FOR ADVICE ON HOW TO SOLVE THIS ISSUE!

Importing the numpy C-extensions failed. This error can happen for
many reasons, often due to issues with your setup or how NumPy was
installed.

We have compiled some common reasons and troubleshooting tips at:

    https://numpy.org/devdocs/user/troubleshooting-importerror.html

Please note and check the following:

  * The Python version is: Python3.10 from "C:\Program Files\cadwork.dir\EXE_30\3d.x64\3D.EXE"
  * The NumPy version is: "1.22.2"

and make sure that they are the versions you expect.
Please carefully study the documentation linked above for further help.

Original error was: No module named 'numpy.core._multiarray_umath'`

**Reply by Brunner246** [reply]
you need to add the path to the package to `sys` . The interpreter needs to know where your `numpy` installation is. 
Be aware that Numpy version 1.22.2 must be used.

```python
sys.path.append("../../site-packages")
```

**Reply by roehligas** [reply]
found the problem on my side... getting a newer Python version in anaconda and spell-checking the file paths did the trick.

**Reply by djdbarroso** [reply]
Hello everyone!
I'm having the same problem here... Included the path to the package but got an error with the compatibility of versions numpy/python, just like @roehligas wrote over.
Does Numpy still have to be version 1.22.2? If so, does it mean that the python version cannot be over 3.10?
Can someone help?

**Reply by NickatLCB** [reply]
If you are positive that pandas is installed, then you are not giving python the correct path to your pandas and numpy packages. This is more than likely your problem.

**Reply by djdbarroso** [reply]
Thanks for your reply, @NickatLCB.
I do have pandas installed and have no problems using it outside Cadwork, in Jupyter Notebook. I guess it must be something with either the path or the versions. I think the problem lies in the versions, but I'm not sure whether I shall "downgrade" Python/pandas to fit into a older version of numpy (1.22.2, like Cadwork aparently demands).

My installed versions are:
Python 3.11.7
Numpy 1.26.4
Pandas 2.1.4

Or, alternatively, do anyone of you know of a module in Python libraries (not external) that can write Excel files?

**Reply by Brunner246** [reply]
@djdbarroso 
In cadwork version 30 (SP 2024) you have to use Python 3.10 and Numpy version 1.22.2

**Reply by djdbarroso** [reply]
I'll try that, then. Thanks, @Brunner246!
